csidriver/pkg/app/controllerserver.go

```diff
@@ -0,0 +1,453 @@
+		klog.Errorf("failed to marshal to string with error: %s", err)
+		return nil, err
+	}
+	klog.Infof("create volume marshal to string: %s", js)
+	msg.Content = js
+	msg.BuildRouter(DefaultReceiveModuleName,
+		GroupResource,
+		resource,
+		constants.CSIOperationTypeCreateVolume)
+
+	// Marshal message
+	reqData, err := json.Marshal(msg)
+	if err != nil {
+		klog.Errorf("marshal request failed with error: %v", err)
+		return nil, err
+	}
+
+	// Send message to KubeEdge
+	resdata, err := send2KubeEdge(string(reqData), cs.keEndpoint)
+	if err != nil {
+		klog.Errorf("send to kubeedge failed with error: %v", err)
+		return nil, err
+	}
+
+	// Unmarshal message
+	result, err := extractMessage(resdata)
+	if err != nil {
+		klog.Errorf("unmarshal response failed with error: %v", err)
+		return nil, err
+	}
+
+	klog.Infof("create volume result: %v", result)
+	data := result.GetContent().(string)
+
+	if msg.GetOperation() == model.ResponseErrorOperation {
+		klog.Errorf("create volume with error: %s", data)
+		return nil, errors.New(data)
+	}
+
+	decodeBytes, err := base64.StdEncoding.DecodeString(data)
+	if err != nil {
+		klog.Errorf("create volume decode with error: %v", err)
+		return nil, err
+	}
+
+	response := &csi.CreateVolumeResponse{}
+	err = json.Unmarshal([]byte(decodeBytes), response)
+	if err != nil {
+		klog.Errorf("create volume unmarshal with error: %v", err)
+		return nil, nil
+	}
+	klog.Infof("create volume response: %v", response)
+
+	createVolumeResponse := &csi.CreateVolumeResponse{}
+	if req.GetVolumeContentSource() != nil {
+		createVolumeResponse = &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				VolumeId:      response.Volume.VolumeId,
+				CapacityBytes: req.GetCapacityRange().GetRequiredBytes(),
+				VolumeContext: req.GetParameters(),
+				ContentSource: req.GetVolumeContentSource(),
+			},
+		}
+	} else {
+		createVolumeResponse = &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				VolumeId:      response.Volume.VolumeId,
+				CapacityBytes: req.GetCapacityRange().GetRequiredBytes(),
+				VolumeContext: req.GetParameters(),
+			},
+		}
+	}
+	return createVolumeResponse, nil
+}
+
+// DeleteVolume issues delete volume func
+func (cs *controllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	// Check arguments
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	// Build message struct
+	msg := model.NewMessage("")
+	resource, err := buildResource(cs.nodeID,
+		DefaultNamespace,
+		constants.CSIResourceTypeVolume,
+		req.GetVolumeId())
+	if err != nil {
+		klog.Errorf("build message resource failed with error: %s", err)
+		return nil, err
+	}
+
+	m := jsonpb.Marshaler{}
+	js, err := m.MarshalToString(req)
+	if err != nil {
+		klog.Errorf("failed to marshal to string with error: %s", err)
+		return nil, err
+	}
+	klog.Infof("delete volume marshal to string: %s", js)
```

Comment:
can we log this with V.(3) or V.(4), too?

Reply:
I will. Thanks